js/websocket_app/src/stream/Data_Stream.tsx

```diff
@@ -0,0 +1,69 @@
+import { useEffect, useState } from 'react';
+import useWebSocket, { ReadyState } from 'react-use-websocket';
+
+function Data_Stream() {
+    const [displayText, setDisplayText] = useState('');
+    const [websocketUrl, setWebsocketUrl] = useState('ws://localhost:8080/ws');
+    const {
+        sendJsonMessage,
+        lastJsonMessage,
+        readyState,
+    } = useWebSocket(websocketUrl, {
+        onOpen: () => console.log('opened'),
+        shouldReconnect: () => true,
+    });
+
+    const connectionStatus = {
+        [ReadyState.CLOSED]: 'Closed',
+        [ReadyState.CLOSING]: 'Closing',
+        [ReadyState.CONNECTING]: 'Connecting',
+        [ReadyState.OPEN]: 'Open',
+        [ReadyState.UNINSTANTIATED]: 'Uninstantiated',
+    }[readyState];
+
+    const handleInputChange = (event: any) => {
+        const inputText = event.target.value;
+        if (/\s$/.test(inputText)) {
+            if (readyState == ReadyState.OPEN) {
+                sendJsonMessage({ text: inputText });
+            }
+        }
+    };
+
+    useEffect(() => {
+        if (lastJsonMessage) {
+            setDisplayText(JSON.stringify(lastJsonMessage, undefined, 2))
+        }
+    }, [lastJsonMessage])
```

Comment:
[Suggestion, duplicated state] `setDisplayText/`displayText` is derivative state of lastJsonMessage`! 

```suggestion
    const displayText = lastJsonMessage ? JSON.stringify(lastJsonMessage, undefined, 2) : "";
```

Reply:
Awesome, done!